tests for rules from the General-Rules package

AngelHely opened the pull request:
This pull request propose tests for rules from the General-Rules package with the tag migrated

AngelHely committed:
beginning test for AsClassRule

AngelHely committed:
test for rule : ReExcessiveVariablesRule and ReInconsistentMethodClassificationRule

AngelHely committed:
Merge branch 'pharo-project:Pharo12' into rulesTest

AngelHely committed:
Test proposition for 5 rules. 

this commit propose 5 tests for rules from Genral-Rules-Tests : ReClassNotCategorizedRule, ReClassVariableCapitalizationRule, ReEquivalentSuperClassMethodsRule, ReExcessiveInheritanceRule, RelongMethodsRuleTest

jecisc commented on src/General-Rules-Tests/ReAsClassRuleTest.class.st:
```diff
@@ -0,0 +1,32 @@
+]
+
+{ #category : 'tests' }
+ReAsClassRuleTest >> testRule [ 
+	|critiques|
+	critiques := self myCritiquesOnMethod: self myMethodCallingAsClassIfPresents.
+	self assert: critiques size equals: 1
```
This one does not work as we have seen but we'll check together how we can fix that.

jecisc commented on src/General-Rules-Tests/ReInconsistentMethodClassificationRuleTest.class.st:
```diff
@@ -0,0 +1,25 @@
+	#tag : 'Migrated'
+}
+
+{ #category : 'accessing' }
+ReInconsistentMethodClassificationRuleTest >> initialize [ 
+]
```
We have a release test that ensure all #initialize methods are in the right protocol and this makes the test fail. 

Instead of having this method here, we could generate the method like this:

```st
self class compile: 'initialize super initialize' classified: 'accessing'.
[ "code of the test" ] ensure: [ (self class >> #initialize) removeFromSystem ]
```

#ensure: will make sure that even if the test crashes, the block deleting the method will be executed.

jecisc commented on src/General-Rules-Tests/ReInconsistentMethodClassificationRuleTest.class.st:
```diff
@@ -0,0 +1,25 @@
+]
+
+{ #category : 'initalize' }
+ReInconsistentMethodClassificationRuleTest >> myCritiques [ 
+^ super myCritiques. 
+]
```
Same comment as above

jecisc commented on src/General-Rules-Tests/ReClassVariableCapitalizationRuleTest.class.st:
```diff
@@ -0,0 +1,34 @@
+]
+
+{ #category : 'tests' }
+ReClassVariableCapitalizationRuleTest >> testPoolVariableNameStartWithLowerCase [
```
It would be nice to add a test in case there is no violation

jecisc commented on src/General-Rules-Tests/ReEquivalentSuperclassMethodsRuleTest.class.st:
```diff
@@ -0,0 +1,26 @@
+]
+
+{ #category : 'tests' }
+ReEquivalentSuperclassMethodsRuleTest >> testRule [
```
It would be nice to add a test in case there is no violation

jecisc commented on src/General-Rules-Tests/ReAsClassRuleTest.class.st:
```diff
@@ -0,0 +1,32 @@
+	#name : 'ReAsClassRuleTest',
+	#superclass : 'ReAbstractRuleTestCase',
+	#classInstVars : [
+		'critiques'
```
This instance variable is unused and will make a release test checking that all instance variables are used fail.

AngelHely committed:
deleete ReAsClassRuleTest

the test for the class ReAsRule is not good, so we delete this test for the moment.

AngelHely committed:
Merge branch 'pharo-project:Pharo12' into rulesTest

AngelHely committed:
Merge branch 'pharo-project:Pharo12' into rulesTest

AngelHely committed:
Merge branch 'pharo-project:Pharo12' into rulesTest

AngelHely committed:
new test methods + improve some tests

add tests in case there is no violation  + improve tests in class ReEquivalentSuperclassMethodsRuleTest + ReInconsistentMethodClassificationRuleTest

jecisc commented on src/General-Rules-Tests/ReInconsistentMethodClassificationRuleTest.class.st:
```diff
@@ -0,0 +1,40 @@
+	self class
+		compile: 'initialize super initialize'
+		classified: 'initialization'.
+	critiques := self myCritiquesOnMethod: self class >> #initialize.
+	self assertEmpty: critiques.
+	[ "code of the test" ] ensure: [
+		(self class >> #initialize) removeFromSystem ]
```
```suggestion
	[ critiques := self myCritiquesOnMethod: self class >> #initialize.
	self assertEmpty: critiques ] ensure: [
		(self class >> #initialize) removeFromSystem ]
```

jecisc commented on src/General-Rules-Tests/ReInconsistentMethodClassificationRuleTest.class.st:
```diff
@@ -0,0 +1,40 @@
+{ #category : 'tests' }
+ReInconsistentMethodClassificationRuleTest >> testRule [
+
+	| critiques |
+	self class
+		compile: 'myCritiques super myCritiques'
+		classified: 'accessing'.
+	critiques := self myCritiquesOnMethod: self class >> #myCritiques.
+	self assert: critiques size equals: 1.
+	[ "code of the test" ] ensure: [
+		(self class >> #myCritiques) removeFromSystem ].
+	self class
+		compile: 'initialize super initialize'
+		classified: 'accessing'.
+	critiques := self myCritiquesOnMethod: self class >> #initialize.
+	self assert: critiques size equals: 1.
+	[ "code of the test" ] ensure: [
```
```suggestion
	| critiques |
	self class
		compile: 'myCritiques super myCritiques'
		classified: 'accessing'.
	[ critiques := self myCritiquesOnMethod: self class >> #myCritiques.
	self assert: critiques size equals: 1 ] ensure: [
		(self class >> #myCritiques) removeFromSystem ].
	self class
		compile: 'initialize super initialize'
		classified: 'accessing'.
	[ critiques := self myCritiquesOnMethod: self class >> #initialize.
	self assert: critiques size equals: 1 ] ensure: [
```

The #ensure: method makes sure the argument blocks gets executed after the block receiving the message, even if the block does not finish (eg. a non local return or an exception happens). So the receiving block must not be empty.

AngelHely committed:
Update src/General-Rules-Tests/ReInconsistentMethodClassificationRuleTest.class.st

Co-authored-by: CyrilFerlicot <cyril@ferlicot.me>

AngelHely committed:
Update src/General-Rules-Tests/ReInconsistentMethodClassificationRuleTest.class.st

Co-authored-by: CyrilFerlicot <cyril@ferlicot.me>
